panic error : Scanning together with the custom template

<!-- 
1. Please search to see if an issue already exists for the bug you encountered.
2. For support requests, FAQs or "How to" questions, please use the GitHub Discussions section instead - https://github.com/projectdiscovery/nuclei/discussions or
3. Join our discord server at https://discord.gg/projectdiscovery and post the question on the #nuclei channel.
-->

<!-- ISSUES MISSING IMPORTANT INFORMATION MAY BE CLOSED WITHOUT INVESTIGATION. -->

### Nuclei version:
```
[ec2-user@ip-10-235-253-152 ~]$ nuclei --version
[INF] Nuclei Engine Version: v3.2.2
[INF] Nuclei Config Directory: /home/ec2-user/.config/nuclei
[INF] Nuclei Cache Directory: /home/ec2-user/.cache/nuclei
[INF] PDCP Directory: /home/ec2-user/.pdcp
```

### Current Behavior: 
[error](panic: runtime error:)


### Expected Behavior:
It seems like the scan should automatically proceed, excluding any templates that cause errors.

### Steps To Reproduce:
I think running it with the custom template causes errors.
If I run only the custom template, there's no error, and there's also no error when running other templates without the custom template...
```
[ec2-user@ip-10-235-253-152 ~]$ /home/ec2-user/go/bin/nuclei -t /home/ec2-user/YeOT/s3/custom_templates/,http/exposed-panels -u google.com

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v3.2.2

		projectdiscovery.io

[WRN] Found 202 templates loaded with deprecated protocol syntax, update before v3 for continued support.
[WRN] Found 4 templates with syntax error (use -validate flag for further examination)
[INF] Current nuclei version: v3.2.2 (latest)
[INF] Current nuclei-templates version: v9.8.0 (latest)
[WRN] Scan results upload to cloud is disabled.
[INF] New templates added in latest release: 85
[INF] Templates loaded for current scan: 1288
[INF] Executing 540 signed templates from projectdiscovery/nuclei-templates
[WRN] Loaded 748 unsigned templates for scan. Use with caution.
[INF] Targets loaded for current scan: 1
[INF] Running httpx on input host
[INF] Found 1 URL from httpx
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
github.com/projectdiscovery/nuclei/v3/pkg/templates.NewClusterExecuter({0xc0014f4cb0, 0x2, 0xc0062057d0?}, 0xc00651ad80)
	/home/ec2-user/go/pkg/mod/github.com/projectdiscovery/nuclei/v3@v3.2.2/pkg/templates/cluster.go:218 +0x846
github.com/projectdiscovery/nuclei/v3/pkg/templates.ClusterTemplates({_, _, _}, {{0x0, 0x0}, {0x0, 0x0}, {{0x0, 0x0}, {{0x0, ...}}, ...}, ...})
	/home/ec2-user/go/pkg/mod/github.com/projectdiscovery/nuclei/v3@v3.2.2/pkg/templates/cluster.go:158 +0x34c
github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).ExecuteScanWithOpts(0xc000736240, {0xc00295e000?, 0x508, 0x550}, {0x35a01d0, 0xc00118d2c0}, 0x0)
	/home/ec2-user/go/pkg/mod/github.com/projectdiscovery/nuclei/v3@v3.2.2/pkg/core/execute_options.go:45 +0x19c
github.com/projectdiscovery/nuclei/v3/internal/runner.(*Runner).executeTemplatesInput(0xc000892000, 0xc000b16240, 0x0?)
	/home/ec2-user/go/pkg/mod/github.com/projectdiscovery/nuclei/v3@v3.2.2/internal/runner/runner.go:633 +0x225
github.com/projectdiscovery/nuclei/v3/internal/runner.(*Runner).runStandardEnumeration(_, {{0x0, 0x0}, {0x0, 0x0}, {{0x0, 0x0}, {{0x0, 0x0}}, {{0x0, ...}}, ...}, ...}, ...)
	/home/ec2-user/go/pkg/mod/github.com/projectdiscovery/nuclei/v3@v3.2.2/internal/runner/runner.go:336 +0x66
github.com/projectdiscovery/nuclei/v3/internal/runner.(*Runner).RunEnumeration(0xc000892000)
	/home/ec2-user/go/pkg/mod/github.com/projectdiscovery/nuclei/v3@v3.2.2/internal/runner/runner.go:548 +0x1628
main.main()
	/home/ec2-user/go/pkg/mod/github.com/projectdiscovery/nuclei/v3@v3.2.2/cmd/nuclei/main.go:170 +0x665
```
When running only the custom template
```
[ec2-user@ip-10-235-253-152 ~]$ /home/ec2-user/go/bin/nuclei -t /home/ec2-user/YeOT/s3/custom_templates/ -u google.com

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v3.2.2

		projectdiscovery.io

[WRN] Found 202 templates loaded with deprecated protocol syntax, update before v3 for continued support.
[WRN] Found 4 templates with syntax error (use -validate flag for further examination)
[INF] Current nuclei version: v3.2.2 (latest)
[INF] Current nuclei-templates version: v9.8.0 (latest)
[WRN] Scan results upload to cloud is disabled.
[INF] New templates added in latest release: 85
[INF] Templates loaded for current scan: 231
[INF] Executing 1 signed templates from projectdiscovery/nuclei-templates
[WRN] Loaded 230 unsigned templates for scan. Use with caution.
[INF] Targets loaded for current scan: 1
[INF] Running httpx on input host
[INF] Found 1 URL from httpx
[INF] Templates clustered: 79 (Reduced 63 Requests)
[report-uri-headers:content_security_policy_report_only] [http] [info] https://google.com [object-src 'none';base-uri 'self';script-src 'nonce-DM6y8rf0R9XAtq1BAxZD1g' 'strict-dynamic' 'report-sample' 'unsafe-eval' 'unsafe-inline' https: http:;report-uri https://csp.withgoogle.com/csp/gws/other-hp]
[report-uri-headers:cross_origin_opener_policy] [http] [info] https://google.com [same-origin-allow-popups; report-to="gws"]
[report-uri-headers:report_to] [http] [info] https://google.com [{"group":"gws","max_age":2592000,"endpoints":[{"url":"https://csp.withgoogle.com/csp/report-to/gws/other"}]}]
[report-uri-headers:x_xss_protection] [http] [info] https://google.com [0]
[report-uri-headers:content_security_policy_report_only] [http] [info] https://google.com [object-src 'none';base-uri 'self';script-src 'nonce-DM6y8rf0R9XAtq1BAxZD1g' 'strict-dynamic' 'report-sample' 'unsafe-eval' 'unsafe-inline' https: http:;report-uri https://csp.withgoogle.com/csp/gws/other-hp]
[report-uri-headers:cross_origin_opener_policy] [http] [info] https://google.com [same-origin-allow-popups; report-to="gws"]
[report-uri-headers:report_to] [http] [info] https://google.com [{"group":"gws","max_age":2592000,"endpoints":[{"url":"https://csp.withgoogle.com/csp/report-to/gws/other"}]}]
[report-uri-headers:x_xss_protection] [http] [info] https://google.com [0]
[INF] Using Interactsh Server: oast.me
[security-txt] [http] [info] https://www.google.com/.well-known/security.txt [https://g.co/vulnz,mailto:security@google.com]
[security-txt] [http] [info] https://www.google.com/.well-known/security.txt [https://g.co/vulnz,mailto:security@google.com]
[security-txt] [http] [info] https://www.google.com/.well-known/security.txt [https://g.co/vulnz,mailto:security@google.com]
...
```
When scanning without the custom template
```
[ec2-user@ip-10-235-253-152 ~]$ /home/ec2-user/go/bin/nuclei -t http/exposed-panels -u google.com

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v3.2.2

		projectdiscovery.io

[INF] Current nuclei version: v3.2.2 (latest)
[INF] Current nuclei-templates version: v9.8.0 (latest)
[WRN] Scan results upload to cloud is disabled.
[INF] New templates added in latest release: 85
[INF] Templates loaded for current scan: 1057
[INF] Executing 539 signed templates from projectdiscovery/nuclei-templates
[WRN] Loaded 518 unsigned templates for scan. Use with caution.
[INF] Targets loaded for current scan: 1
[INF] Running httpx on input host
[INF] Found 1 URL from httpx
[INF] Templates clustered: 507 (Reduced 479 Requests)
[INF] No results found. Better luck next time!
```


### Anything else:

If you need any additional information, feel free to let me know.

--- a/pkg/protocols/http/request.go
+++ b/pkg/protocols/http/request.go
@@ -394,6 +394,8 @@ func (request *Request) ExecuteWithResults(input *contextargs.Context, dynamicVa
 			}
 			var gotMatches bool
 			execReqErr := request.executeRequest(input, generatedHttpRequest, previous, hasInteractMatchers, func(event *output.InternalWrappedEvent) {
+				gologger.Error().Msgf("event is %+v", event)
+				gologger.Error().Msgf("internal event is %+v", event.InternalEvent)
 				// a special case where operators has interactsh matchers and multiple request are made
 				// ex: status_code_2 , interactsh_protocol (from 1st request) etc
 				needsRequestEvent := interactsh.HasMatchers(request.CompiledOperators) && request.NeedsRequestCondition()
@@ -472,18 +474,16 @@ func (request *Request) ExecuteWithResults(input *contextargs.Context, dynamicVa
 const drainReqSize = int64(8 * 1024)
 
 // executeRequest executes the actual generated request and returns error if occurred
-func (request *Request) executeRequest(input *contextargs.Context, generatedRequest *generatedRequest, previousEvent output.InternalEvent, hasInteractMatchers bool, callback protocols.OutputEventCallback, requestCount int) (err error) {
-	var event *output.InternalWrappedEvent
-	defer func() {
-		if event != nil {
-			if event.InternalEvent == nil {
-				event.InternalEvent = make(map[string]interface{})
-				event.InternalEvent["template-id"] = request.options.TemplateID
-			}
-			// add the request URL pattern to the event
-			event.InternalEvent[ReqURLPatternKey] = generatedRequest.requestURLPattern
-		}
-	}()
+func (request *Request) executeRequest(input *contextargs.Context, generatedRequest *generatedRequest, previousEvent output.InternalEvent, hasInteractMatchers bool, processEvent protocols.OutputEventCallback, requestCount int) (err error) {
+
+	// wrap one more callback for validation and fixing event
+	callback := func(event *output.InternalWrappedEvent) {
+		// validateNFixEvent performs necessary validation on generated event
+		// and attempts to fix it , this includes things like making sure
+		// `template-id` is set , `request-url-pattern` is set etc
+		request.validateNFixEvent(generatedRequest, event)
+		processEvent(event)
+	}
 
 	request.setCustomHeaders(generatedRequest)
 
@@ -692,7 +692,7 @@ func (request *Request) executeRequest(input *contextargs.Context, generatedRequ
 		if len(generatedRequest.interactshURLs) > 0 {
 			// according to logic we only need to trigger a callback if interactsh was used
 			// and request failed in hope that later on oast interaction will be received
-			event = &output.InternalWrappedEvent{}
+			event := &output.InternalWrappedEvent{}
 			if request.CompiledOperators != nil && request.CompiledOperators.HasDSL() {
 				event.InternalEvent = outputEvent
 			}
@@ -807,7 +807,7 @@ func (request *Request) executeRequest(input *contextargs.Context, generatedRequ
 		// prune signature internal values if any
 		request.pruneSignatureInternalValues(generatedRequest.meta)
 
-		event = eventcreator.CreateEventWithAdditionalOptions(request, generators.MergeMaps(generatedRequest.dynamicValues, finalEvent), request.options.Options.Debug || request.options.Options.DebugResponse, func(internalWrappedEvent *output.InternalWrappedEvent) {
+		event := eventcreator.CreateEventWithAdditionalOptions(request, generators.MergeMaps(generatedRequest.dynamicValues, finalEvent), request.options.Options.Debug || request.options.Options.DebugResponse, func(internalWrappedEvent *output.InternalWrappedEvent) {
 			internalWrappedEvent.OperatorsResult.PayloadValues = generatedRequest.meta
 		})
 		if hasInteractMatchers {
@@ -841,6 +841,19 @@ func (request *Request) executeRequest(input *contextargs.Context, generatedRequ
 	}
 	// return project file save error if any
 	return errx
+}
+
+// validateNFixEvent validates and fixes the event
+// it adds any missing template-id and request-url-pattern
+func (request *Request) validateNFixEvent(gr *generatedRequest, event *output.InternalWrappedEvent) {
+	if event != nil {
+		if event.InternalEvent == nil {
+			event.InternalEvent = make(map[string]interface{})
+			event.InternalEvent["template-id"] = request.options.TemplateID
+		}
+		// add the request URL pattern to the event
+		event.InternalEvent[ReqURLPatternKey] = gr.requestURLPattern
+	}
 }
 
 // handleSignature of the http request

--- a/pkg/protocols/http/request.go
+++ b/pkg/protocols/http/request.go
@@ -394,8 +394,6 @@ func (request *Request) ExecuteWithResults(input *contextargs.Context, dynamicVa
 			}
 			var gotMatches bool
 			execReqErr := request.executeRequest(input, generatedHttpRequest, previous, hasInteractMatchers, func(event *output.InternalWrappedEvent) {
-				gologger.Error().Msgf("event is %+v", event)
-				gologger.Error().Msgf("internal event is %+v", event.InternalEvent)
 				// a special case where operators has interactsh matchers and multiple request are made
 				// ex: status_code_2 , interactsh_protocol (from 1st request) etc
 				needsRequestEvent := interactsh.HasMatchers(request.CompiledOperators) && request.NeedsRequestCondition()
@@ -481,7 +479,7 @@ func (request *Request) executeRequest(input *contextargs.Context, generatedRequ
 		// validateNFixEvent performs necessary validation on generated event
 		// and attempts to fix it , this includes things like making sure
 		// `template-id` is set , `request-url-pattern` is set etc
-		request.validateNFixEvent(generatedRequest, event)
+		request.validateNFixEvent(input, generatedRequest, err, event)
 		processEvent(event)
 	}
 
@@ -845,7 +843,7 @@ func (request *Request) executeRequest(input *contextargs.Context, generatedRequ
 
 // validateNFixEvent validates and fixes the event
 // it adds any missing template-id and request-url-pattern
-func (request *Request) validateNFixEvent(gr *generatedRequest, event *output.InternalWrappedEvent) {
+func (request *Request) validateNFixEvent(input *contextargs.Context, gr *generatedRequest, err error, event *output.InternalWrappedEvent) {
 	if event != nil {
 		if event.InternalEvent == nil {
 			event.InternalEvent = make(map[string]interface{})
@@ -853,6 +851,24 @@ func (request *Request) validateNFixEvent(gr *generatedRequest, event *output.In
 		}
 		// add the request URL pattern to the event
 		event.InternalEvent[ReqURLPatternKey] = gr.requestURLPattern
+		if event.InternalEvent["host"] == nil {
+			event.InternalEvent["host"] = input.MetaInput.Input
+		}
+		if event.InternalEvent["template-id"] == nil {
+			event.InternalEvent["template-id"] = request.options.TemplateID
+		}
+		if event.InternalEvent["type"] == nil {
+			event.InternalEvent["type"] = request.Type().String()
+		}
+		if event.InternalEvent["template-path"] == nil {
+			event.InternalEvent["template-path"] = request.options.TemplatePath
+		}
+		if event.InternalEvent["template-info"] == nil {
+			event.InternalEvent["template-info"] = request.options.TemplateInfo
+		}
+		if err != nil {
+			event.InternalEvent["error"] = err.Error()
+		}
 	}
 }
 

--- a/pkg/templates/cluster.go
+++ b/pkg/templates/cluster.go
@@ -59,6 +59,14 @@ func Cluster(list []*Template) [][]*Template {
 			final = append(final, []*Template{template})
 			continue
 		}
+
+		// it is not possible to cluster flow and multiprotocol due to dependent execution
+		if template.Flow != "" || template.Options.IsMultiProtocol {
+			_ = skip.Set(key, struct{}{})
+			final = append(final, []*Template{template})
+			continue
+		}
+
 		_ = skip.Set(key, struct{}{})
 
 		var templateType types.ProtocolType
@@ -78,6 +86,13 @@ func Cluster(list []*Template) [][]*Template {
 			otherKey := other.Path
 
 			if skip.Has(otherKey) {
+				continue
+			}
+
+			// it is not possible to cluster flow and multiprotocol due to dependent execution
+			if other.Flow != "" || other.Options.IsMultiProtocol {
+				_ = skip.Set(otherKey, struct{}{})
+				final = append(final, []*Template{other})
 				continue
 			}
 

--- a/pkg/templates/cluster_test.go
+++ b/pkg/templates/cluster_test.go
@@ -3,15 +3,25 @@ package templates
 import (
 	"testing"
 
+	"github.com/projectdiscovery/nuclei/v3/pkg/model"
+	"github.com/projectdiscovery/nuclei/v3/pkg/model/types/severity"
 	"github.com/projectdiscovery/nuclei/v3/pkg/protocols/dns"
 	"github.com/projectdiscovery/nuclei/v3/pkg/protocols/http"
+	"github.com/projectdiscovery/nuclei/v3/pkg/testutils"
 	"github.com/stretchr/testify/require"
 )
 
 func TestClusterTemplates(t *testing.T) {
+	// state of whether template is flow or multiprotocol is stored in executerOptions i.e why we need to pass it
+	execOptions := testutils.NewMockExecuterOptions(testutils.DefaultOptions, &testutils.TemplateInfo{
+		ID:   "templateID",
+		Info: model.Info{SeverityHolder: severity.Holder{Severity: severity.Low}, Name: "test"},
+	})
 	t.Run("http-cluster-get", func(t *testing.T) {
 		tp1 := &Template{Path: "first.yaml", RequestsHTTP: []*http.Request{{Path: []string{"{{BaseURL}}"}}}}
 		tp2 := &Template{Path: "second.yaml", RequestsHTTP: []*http.Request{{Path: []string{"{{BaseURL}}"}}}}
+		tp1.Options = execOptions
+		tp2.Options = execOptions
 		tpls := []*Template{tp1, tp2}
 		// cluster 0
 		expected := []*Template{tp1, tp2}
@@ -21,6 +31,8 @@ func TestClusterTemplates(t *testing.T) {
 	t.Run("no-http-cluster", func(t *testing.T) {
 		tp1 := &Template{Path: "first.yaml", RequestsHTTP: []*http.Request{{Path: []string{"{{BaseURL}}/random"}}}}
 		tp2 := &Template{Path: "second.yaml", RequestsHTTP: []*http.Request{{Path: []string{"{{BaseURL}}/another"}}}}
+		tp1.Options = execOptions
+		tp2.Options = execOptions
 		tpls := []*Template{tp1, tp2}
 		expected := [][]*Template{{tp1}, {tp2}}
 		got := Cluster(tpls)
@@ -29,6 +41,8 @@ func TestClusterTemplates(t *testing.T) {
 	t.Run("dns-cluster", func(t *testing.T) {
 		tp1 := &Template{Path: "first.yaml", RequestsDNS: []*dns.Request{{Name: "{{Hostname}}"}}}
 		tp2 := &Template{Path: "second.yaml", RequestsDNS: []*dns.Request{{Name: "{{Hostname}}"}}}
+		tp1.Options = execOptions
+		tp2.Options = execOptions
 		tpls := []*Template{tp1, tp2}
 		// cluster 0
 		expected := []*Template{tp1, tp2}

--- a/pkg/tmplexec/exec.go
+++ b/pkg/tmplexec/exec.go
@@ -92,7 +92,11 @@ func (e *TemplateExecuter) Requests() int {
 
 // Execute executes the protocol group and returns true or false if results were found.
 func (e *TemplateExecuter) Execute(ctx *scan.ScanContext) (bool, error) {
-	results := &atomic.Bool{}
+	// executed contains status of execution if it was successfully executed or not
+	// doesn't matter if it was matched or not
+	executed := &atomic.Bool{}
+	// matched in this case means something was exported / written to output
+	matched := &atomic.Bool{}
 	defer func() {
 		// it is essential to remove template context of `Scan i.e template x input pair`
 		// since it is of no use after scan is completed (regardless of success or failure)
@@ -101,11 +105,11 @@ func (e *TemplateExecuter) Execute(ctx *scan.ScanContext) (bool, error) {
 
 	var lastMatcherEvent *output.InternalWrappedEvent
 	writeFailureCallback := func(event *output.InternalWrappedEvent, matcherStatus bool) {
-		if !results.Load() && matcherStatus {
+		if !matched.Load() && matcherStatus {
 			if err := e.options.Output.WriteFailure(event); err != nil {
 				gologger.Warning().Msgf("Could not write failure event to output: %s\n", err)
 			}
-			results.CompareAndSwap(false, true)
+			executed.CompareAndSwap(false, true)
 		}
 	}
 
@@ -133,11 +137,11 @@ func (e *TemplateExecuter) Execute(ctx *scan.ScanContext) (bool, error) {
 		// If no results were found, and also interactsh is not being used
 		// in that case we can skip it, otherwise we've to show failure in
 		// case of matcher-status flag.
-		if !event.HasOperatorResult() && !event.UsesInteractsh {
+		if !event.HasOperatorResult() && event.InternalEvent != nil {
 			lastMatcherEvent = event
 		} else {
 			if writer.WriteResult(event, e.options.Output, e.options.Progress, e.options.IssuesClient) {
-				results.CompareAndSwap(false, true)
+				matched.Store(true)
 			} else {
 				lastMatcherEvent = event
 			}
@@ -152,7 +156,7 @@ func (e *TemplateExecuter) Execute(ctx *scan.ScanContext) (bool, error) {
 	// so in compile step earlier we compile it to validate javascript syntax and other things
 	// and while executing we create new instance of flow executor everytime
 	if e.options.Flow != "" {
-		flowexec, err := flow.NewFlowExecutor(e.requests, ctx, e.options, results, e.program)
+		flowexec, err := flow.NewFlowExecutor(e.requests, ctx, e.options, executed, e.program)
 		if err != nil {
 			ctx.LogError(err)
 			return false, fmt.Errorf("could not create flow executor: %s", err)
@@ -169,7 +173,7 @@ func (e *TemplateExecuter) Execute(ctx *scan.ScanContext) (bool, error) {
 	if lastMatcherEvent != nil {
 		writeFailureCallback(lastMatcherEvent, e.options.Options.MatcherStatus)
 	}
-	return results.Load(), errx
+	return executed.Load(), errx
 }
 
 // ExecuteWithResults executes the protocol requests and returns results instead of writing them.

--- a/pkg/tmplexec/exec.go
+++ b/pkg/tmplexec/exec.go
@@ -173,7 +173,7 @@ func (e *TemplateExecuter) Execute(ctx *scan.ScanContext) (bool, error) {
 	if lastMatcherEvent != nil {
 		writeFailureCallback(lastMatcherEvent, e.options.Options.MatcherStatus)
 	}
-	return executed.Load(), errx
+	return executed.Load() || matched.Load(), errx
 }
 
 // ExecuteWithResults executes the protocol requests and returns results instead of writing them.

